prototype/FFT_image_processing/bitmap.c

```diff
@@ -0,0 +1,268 @@
+      + imageData[i * RGB_BYTES + 1] * wGreen
+      + imageData[i * RGB_BYTES + 2] * wRed;
+      greyData[i] = (unsigned char) y; 
+  }
+
+  return greyData;
+
+}
+
+/**
+ * @brief Reads a bitmap file
+ * @param filename The bitmap filename with the extension
+ * @param bitmapInfoHeader The struct that will hold the BITMAPINFOHEADER
+ * of the opened bitmap file
+ * @return The entire image data, in bytes. This data must be freed after use to
+ * prevent memory leaks
+*/
+unsigned char* readBitmapImage(char *filename, BITMAPINFOHEADER *bitmapInfoHeader) {
+  
+  FILE *imageFile;
+  BITMAPFILEHEADER bitmapFileHeader;
+
+  //open file in read binary mode
+  imageFile = fopen(filename,"rb");
+  if (imageFile == NULL) {
+    return NULL;
+  }
+
+  // Read individual bitmap file header properties
+  // Must be read individually to prevent packing issues
+  fread(&bitmapFileHeader.bfType, sizeof(bitmapFileHeader.bfType), 1, imageFile);
+  fread(&bitmapFileHeader.bfSize, sizeof(bitmapFileHeader.bfSize), 1, imageFile);
+  fread(&bitmapFileHeader.bfReserved1, sizeof(bitmapFileHeader.bfReserved1), 1, imageFile);
+  fread(&bitmapFileHeader.bfReserved2, sizeof(bitmapFileHeader.bfReserved2), 1, imageFile);
+  fread(&bitmapFileHeader.bfOffBits, sizeof(bitmapFileHeader.bfOffBits), 1, imageFile);
+
+  // Verify for the bm file type
+  if (bitmapFileHeader.bfType != 0x4D42)
+  {
+      fclose(imageFile);
+      return NULL;
+  }
+
+  // Read individual bitmap file header properties
+  // Must be read individually to prevent packing issues
+  fread(&bitmapInfoHeader->biSize, sizeof(bitmapInfoHeader->biSize), 1, imageFile);
+  fread(&bitmapInfoHeader->biWidth, sizeof(bitmapInfoHeader->biWidth), 1, imageFile);
+  fread(&bitmapInfoHeader->biHeight, sizeof(bitmapInfoHeader->biHeight), 1, imageFile);
+  fread(&bitmapInfoHeader->biPlanes, sizeof(bitmapInfoHeader->biPlanes), 1, imageFile);
+  fread(&bitmapInfoHeader->biBitCount, sizeof(bitmapInfoHeader->biBitCount), 1, imageFile);
+  fread(&bitmapInfoHeader->biCompression, sizeof(bitmapInfoHeader->biCompression), 1, imageFile);
+  fread(&bitmapInfoHeader->biSizeImage, sizeof(bitmapInfoHeader->biSizeImage), 1, imageFile);
+  fread(&bitmapInfoHeader->biXPelsPerMeter, sizeof(bitmapInfoHeader->biXPelsPerMeter), 1, imageFile);
+  fread(&bitmapInfoHeader->biYPelsPerMeter, sizeof(bitmapInfoHeader->biYPelsPerMeter), 1, imageFile);
+  fread(&bitmapInfoHeader->biClrUsed, sizeof(bitmapInfoHeader->biClrUsed), 1, imageFile);
+  fread(&bitmapInfoHeader->biClrImportant, sizeof(bitmapInfoHeader->biClrImportant), 1, imageFile);
+
+  // Go where the data is
+  fseek(imageFile, bitmapFileHeader.bfOffBits, SEEK_SET);
+
+  // handle cases where the size is 0
+  if (bitmapInfoHeader->biSizeImage == 0) {
+    bitmapInfoHeader->biSizeImage = bitmapFileHeader.bfSize - bitmapFileHeader.bfOffBits;
+  }
+
+  // Allocate memory for the bitmap data
+  unsigned char* imageData = (unsigned char*)malloc(bitmapInfoHeader->biSizeImage);
```

Comment:
Again, malloc but no free? I understand  why you're doing that and I think it's fair in that case. But, I think you should either document the fact that the user is responsible for freeing the memory or give a method similar to how the File object works.

Reply:
A comment was added